### PR TITLE
Kubernetes Caching Watchers Implementation

### PR DESCRIPTION
**Pull request by mbolt35**
Use caching watchers to maintain up to date cached cluster resources

### CODE REVIEW

File: costmodel/watchcontroller.go [hunk 1]
```diff
@@ -2,34 +2,104 @@ package costmodel
...
+
+	// GetAll returns all of the resources
+	GetAll() []interface{}
+
+	// SetUpdateHandler sets a specific handler for adding/updating individual resources
+	SetUpdateHandler(WatchHandler) WatchController
+
+	// SetRemovedHandler sets a specific handler for removing individual resources
+	SetRemovedHandler(WatchHandler) WatchController
+}
```
Comment (mbolt35) [inline costmodel/watchcontroller.go]: Added the above APIs in order to specifically watch a resource, but these aren't exposed via the `CacheCluster` interface yet. We'll want to hook these in when we start listening to pod updates.

File: costmodel/clustercache.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+func (kcc *KubernetesClusterCache) Run(stopCh chan struct{}) {
+	go kcc.namespaceWatch.Run(1, stopCh)
+	go kcc.nodeWatch.Run(1, stopCh)
+	go kcc.podWatch.Run(1, stopCh)
+	go kcc.serviceWatch.Run(1, stopCh)
+	go kcc.deploymentsWatch.Run(1, stopCh)
+	go kcc.pvWatch.Run(1, stopCh)
+	go kcc.storageClassWatch.Run(1, stopCh)
+}
```
Comment (mbolt35) [inline costmodel/clustercache.go]: I'm pretty sure it's ok to use the same `stopCh` here as long as we actually stop using `close(stopCh)`

File: costmodel/costmodel.go [hunk 3]
```diff
@@ -44,52 +42,20 @@ const (
...
-	cm.ContainerWatcher(podListWatcher)
-	return cm
+	stop chan struct{}
 }
 
-func (cm *CostModel) ContainerWatcher(podListWatcher cache.ListerWatcher) {
-
-	// create the workqueue
-	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+func NewCostModel(client kubernetes.Interface) *CostModel {
```
Comment (AjayTripathy) [inline costmodel/costmodel.go]: The underlying controllers should provide a HasSynced() function for an "initial sync" of your underlying caches. I'd prefer if this waited to return until all the controllers have synced. No need to do anything fancy, I think a while loop in KubernetesClusterCache should do the trick.

File: costmodel/watchcontroller.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+)
+
+// Type alias for a receiver func
+type WatchHandler = func(interface{})
+
+// WatchController defines a contract for an object which watches a specific resource set for
+// add, updates, and removals
+type WatchController interface {
+	// Initializes the cache
+	WarmUp(chan struct{})
```
Comment (mbolt35) [inline costmodel/watchcontroller.go]: Split the Run() into WarmUp() and Run() where WarmUp waits for the HasSync on the informer.

File: costmodel/clustercache.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+	go initializeCache(kcc.namespaceWatch, &wg, cancel)
+	go initializeCache(kcc.nodeWatch, &wg, cancel)
+	go initializeCache(kcc.podWatch, &wg, cancel)
+	go initializeCache(kcc.serviceWatch, &wg, cancel)
+	go initializeCache(kcc.deploymentsWatch, &wg, cancel)
+	go initializeCache(kcc.pvWatch, &wg, cancel)
+	go initializeCache(kcc.storageClassWatch, &wg, cancel)
+
+	wg.Wait()
```
Comment (mbolt35) [inline costmodel/clustercache.go]: Executes blocking WarmUp() in a go func for each of the watchers. WarmUp waits for HasSync on the informer, so this should return the cluster cache populated.

**Pull request merged**